### PR TITLE
intern: set finalizer before storing uintptr in valMap

### PR DESCRIPTION
**Pull request by mdlayher**
Signed-off-by: Matt Layher <mdlayher@gmail.com>

Updates #13.

### CODE REVIEW

File: intern.go [hunk 1]
```diff
@@ -117,8 +117,13 @@ func get(k key) *Value {
 	if valSafe != nil {
 		valSafe[k] = v
 	} else {
-		valMap[k] = uintptr(unsafe.Pointer(v))
+		// A note from bcmills: "... the uintptr conversion should happen after
+		// the call to runtime.SetFinalizer, not before. Even with the current
+		// (as of Go 1.15) memory model, the Value could be allocated on the
+		// stack before the finalizer is set, then moved to the heap when the
```
Comment (josharian) [inline intern.go:123]: If v is allocated on the stack, runtime.SetFinalizer might fail (as opposed to moving v to the stack). We could fashion some other means to force v to escape first. But I'm not entirely convinced it makes sense to complicate the code now to protect against complex hypothetical compiler optimizations.

Comment (mdlayher) [reply intern.go:123]: TIL. Would you prefer I leave the code as-is but add a quick comment noting the potential footgun described in #13? Or do you think this should just be closed out?

Comment (josharian) [reply intern.go:123]: I'm inclined to add this note from Bryan to the long "implementation details and dangers" comment near the end of the file. But this change doesn't do any harm, either. Let's wait for @bradfitz or @danderson to weigh in.

Thanks for following up on this!

Comment (bradfitz) [reply intern.go:123]: If you want to submit this, at least massively reduce this comment to something like:

```go
// call SetFinalizer before uintptr conversion (theoretical concern;
// see https://github.com/go4org/intern/issues/13)
```

**Comment by mdlayher** [comment]
@bradfitz done, PTAL and merge when ready.

**Review by bradfitz — APPROVED** [review]

**Pull request merged**